Comprehensive selectors

JackDanger opened the pull request:
@jaykz52 This is a complete implementation of CSS-style selectors. I added tests and updated the README.

We're using this at Square already and we need the single-string selector because it's much easier to store away for retry logic than trying to store a selector + a context it's run inside of.

Thanks for all your great work!

JackDanger committed:
extending the UIA shim

JackDanger committed:
Adding specs for comprehensive selectors

JackDanger committed:
functional, needs refactoring

JackDanger committed:
integrated E_filter

JackDanger committed:
finished integrating comprehensive selectors

Tests are still green :)

JackDanger committed:
updating README

jaykz52 commented:
@JackDanger I'm really excited to dig in. Will try to start diving in tonight. Glad to here there's some tests around this (I really need to invest time into some sort of Test App that can be used for asserting against real-live UIAutomation rather than a shim).

I'm working with the QA team at Twitter to start to leverage mechanic internally as well. If you'd like to grab coffee/beer sometime, I'd appreciate hearing your input on what's working and what's not, and what still needs to be done.

JackDanger commented:
Cool! The tests are pretty exhaustive but if you'd prefer I break it out into new sections of tests I'd be happy to. The changes actually aren't that radical, it's just filling in the missing features.

Interestingly, a side-benefit of the work I did for selectors fixed the $.fn.filter function so it works around Apple's broken referential transparency.

I'd love to stop by the Twitter office (I've still never been!) to talk about this stuff. We should set something up.

JackDanger committed:
whoops, forgot this console.log line

JackDanger committed:
removing two unused vars

jaykz52 commented:
No need to break out. I was just saying that I'm glad to see additional tests, as my prior test coverage around the selecting logic was minimal.

JackDanger committed:
fixing bug with name matching

Commit:
adding selecting by any attribute, not just name

Before:

   [name=Some Thing]

After:

   [name=Some Thing] or
   [label=Some Thing] or
   [value=Some Thing]

I need this so I can quickly match whatever text field happens to be the
one displaying 'First Name' on my signup form. So this now works:

  $("textfield[value=First Name]")

Commit:
adding reference to searching by value to README
